RecoLocalCalo/HGCalRecProducers/plugins/KernelManagerHGCalRecHit.cu

```diff
@@ -0,0 +1,59 @@
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <inttypes.h>
+#include "RecoLocalCalo/HGCalRecProducers/plugins/KernelManagerHGCalRecHit.h"
+#include "RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitKernelImpl.cuh"
+
+KernelManagerHGCalRecHit::KernelManagerHGCalRecHit(HGCUncalibratedRecHitSoA* h_uncalibSoA,
+                                                   HGCUncalibratedRecHitSoA* d_uncalibSoA,
+                                                   HGCRecHitSoA* d_calibSoA)
+    : h_uncalibSoA_(h_uncalibSoA), d_uncalibSoA_(d_uncalibSoA), d_calibSoA_(d_calibSoA) {
+  nhits_ = h_uncalibSoA_->nhits_;
+  stride_ = h_uncalibSoA_->stride_;
+  ::nb_rechits_ = (stride_ + ::nt_rechits_.x - 1) / ::nt_rechits_.x;
+  nbytes_device_ = d_uncalibSoA_->nbytes_ * stride_;
+}
+
+KernelManagerHGCalRecHit::KernelManagerHGCalRecHit(HGCRecHitSoA* h_calibSoA, HGCRecHitSoA* d_calibSoA)
+    : h_calibSoA_(h_calibSoA), d_calibSoA_(d_calibSoA) {
+  nhits_ = h_calibSoA_->nhits_;
+  stride_ = h_calibSoA_->stride_;
+  ::nb_rechits_ = (stride_ + ::nt_rechits_.x - 1) / ::nt_rechits_.x;
+  nbytes_host_ = h_calibSoA_->nbytes_ * stride_;
+}
+
+KernelManagerHGCalRecHit::~KernelManagerHGCalRecHit() {}
+
+void KernelManagerHGCalRecHit::transfer_soa_to_device_(const cudaStream_t& stream) {
+  cudaCheck(cudaMemcpyAsync(
+      d_uncalibSoA_->amplitude_, h_uncalibSoA_->amplitude_, nbytes_device_, cudaMemcpyHostToDevice, stream));
+  cudaCheck(cudaGetLastError());
```

Comment:
IIUC `cudaGetLastError()` is not very useful after `cudaMemcpyAsync()`.

Reply:
As described [here](http://horacio9573.no-ip.org/cuda/group__CUDART__MEMORY_g732efed5ab5cb184c920a21eb36e8ce4.html), it seems the returned ```cudaError_t``` value has a meaning even when the function returns before the copy is finalized. When the copy is done before the method returns, it behaves exactly as [cudaMemcpy](http://horacio9573.no-ip.org/cuda/group__CUDART__MEMORY_g48efa06b81cc031b2aa6fdc2e9930741.html). In both scenarios it seems the method is useful.
Let me know if there is any canonical way of handling CUDA errors within CMSSW, or if these calls are somehow "slow" and should be avoided.

Reply:
`cudaMemcpyAsync()` itself returns `cudaError_t` about launching the asynchronous memory copy. In case of failures in the asynchronous operation after the launch, if I've understood correctly, nearly all CUDA API calls (can) return the error code for that error in asynchronous operation. If the job is in the state of waiting one asynchronous CUDA operation to complete, and that fails, we get the error code for that in the callback function
https://github.com/cms-sw/cmssw/blob/188ed7abcf7bac9cf0b09471061a7ee2ba87373e/HeterogeneousCore/CUDACore/src/ScopedContext.cc#L17

To me `cudaGetLastError()`is mostly useful to check if kernel launch fails, because the `kernel<<<...>>>();` syntax does not allow any other way to check (in contrast to API function calls). Elsewhere it is mostly confusing ("why is this here?"), even if the runtime cost may be negligible.